File: nup/src/main/java/PlaybackReporter.java
package org.erat.nup;

import android.util.Log;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.Date;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class PlaybackReporter {
    private static final String TAG = "PlaybackReporter";

    private final SongDatabase songDb;
    private final Downloader downloader;
    private final TaskRunner taskRunner;
    private final NetworkHelper networkHelper;

    private final ReentrantLock lock = new ReentrantLock();

    public PlaybackReporter(
            SongDatabase songDb,
            Downloader downloader,
            TaskRunner taskRunner,
            NetworkHelper networkHelper) {
        this.songDb = songDb;
        this.downloader = downloader;
        this.taskRunner = taskRunner;
        this.networkHelper = networkHelper;

        // TODO: Listen for the network coming up and send pending reports then?

        // Retry all of the pending reports in the background.
        if (networkHelper.isNetworkAvailable()) {
            this.taskRunner.runInBackground(
                    new Runnable() {
                        @Override
                        public void run() {
                            lock.lock();
                            try {
                                reportPendingSongs();
                            } finally {
                                lock.unlock();
                            }
                        }
                    });
        }
    }

    /**
     * Asynchronously reports the playback of a song.
     *
     * @param songId ID of played song
     * @param startDate time at which playback started
     */
    public void report(final long songId, final Date startDate) {
        taskRunner.runInBackground(
                new Runnable() {
                    @Override
                    public void run() {
                        lock.lock();
                        try {
                            songDb.addPendingPlaybackReport(songId, startDate);
                            if (networkHelper.isNetworkAvailable()) {
                                reportPendingSongs();
                            }
                        } finally {
                            lock.unlock();
                        }
                    }
                });
    }

    /** Reports all unreported songs. */
    public void reportPendingSongs() {
        if (!lock.isHeldByCurrentThread()) {
            throw new IllegalStateException("Lock not held");
        }

        List<SongDatabase.PendingPlaybackReport> reports = songDb.getAllPendingPlaybackReports();
        for (SongDatabase.PendingPlaybackReport report : reports) {
            if (reportInternal(report.songId, report.startDate)) {
                songDb.removePendingPlaybackReport(report.songId, report.startDate);
            }
        }
    }

    /**
     * Synchronously reports the playback of a song.
     *
     * @param songId ID of played song
     * @param startDate time at which playback started
     * @return <code>true</code> if the report was successfully received by the server
     */
    private boolean reportInternal(long songId, Date startDate) {
        if (!networkHelper.isNetworkAvailable()) {
            return false;
        }

        Log.d(TAG, "reporting song " + songId + " started at " + startDate);
        HttpURLConnection conn = null;
        try {
            String path =
                    String.format(
                            "/report_played?songId=%d&startTime=%f",
                            songId, startDate.getTime() / 1000.0);
            conn =
                    downloader.download(
                            downloader.getServerUrl(path),
                            "POST",
                            Downloader.AuthType.SERVER,
                            null);
            if (conn.getResponseCode() == 200) {
                return true;
            }
            Log.e(
                    TAG,
                    "got " + conn.getResponseCode() + " from server: " + conn.getResponseMessage());
        } catch (Downloader.PrefException e) {
            Log.e(TAG, "got preferences error: " + e);
        } catch (IOException e) {
            Log.e(TAG, "got IO error: " + e);
        } finally {
            if (conn != null) {
                conn.disconnect();
            }
        }
        return false;
    }
}

File: nup/src/main/java/PlaybackReporter.kt
/*
 * Copyright 2011 Daniel Erat <dan@erat.org>
 * All rights reserved.
 */

package org.erat.nup

import android.util.Log
import java.io.IOException
import java.net.HttpURLConnection
import java.util.Date
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.erat.nup.Downloader.PrefException

class PlaybackReporter(
    private val scope: CoroutineScope,
    private val songDb: SongDatabase,
    private val downloader: Downloader,
    private val networkHelper: NetworkHelper,
    private val pendingDispatcher: CoroutineDispatcher = Dispatchers.IO, // for tests
) : NetworkHelper.Listener {
    private val lock = ReentrantLock()
    private var reporting = false

    /** Start listening for network changes and send pending reports if possible. */
    fun start() {
        networkHelper.addListener(this)
        if (networkHelper.isNetworkAvailable) scope.launch(pendingDispatcher) { reportPending() }
    }

    override fun onNetworkAvailabilityChange(available: Boolean) {
        if (available) scope.launch(pendingDispatcher) { reportPending() }
    }

    /** Report the playback of [songId] starting at [startDate]. */
    suspend fun report(songId: Long, startDate: Date) {
        songDb.addPendingPlaybackReport(songId, startDate)
        if (networkHelper.isNetworkAvailable) reportPending()
    }

    /** Synchronously report all unreported songs. */
    suspend fun reportPending() {
        lock.withLock() {
            if (reporting) return
            reporting = true
        }

        val reports = songDb.allPendingPlaybackReports()
        if (!reports.isEmpty()) Log.d(TAG, "Sending ${reports.size} pending report(s)")
        for (report in reports) {
            if (send(report.songId, report.startDate)) {
                songDb.removePendingPlaybackReport(report.songId, report.startDate)
            }
        }

        // report() could've added more songs while we were sending the pending reports to the
        // server. This probably isn't a big deal, since we'll send those songs next time. Kotlin
        // won't let us call suspend functions in critical sections.
        lock.withLock() { reporting = false }
    }

    /** Send a report of [songId] starting at [startDate]. */
    private suspend fun send(songId: Long, startDate: Date): Boolean {
        if (!networkHelper.isNetworkAvailable) return false

        Log.d(TAG, "Reporting song $songId started at $startDate")
        var conn: HttpURLConnection? = null
        try {
            val start = String.format("%.3f", startDate.time / 1000.0)
            val path = "/played?songId=$songId&startTime=$start"
            conn = downloader.download(
                downloader.getServerUrl(path),
                "POST",
                Downloader.AuthType.SERVER,
                null
            )
            if (conn.responseCode == 200) return true
            Log.e(TAG, "Got ${conn.responseCode} from server: ${conn.responseMessage}")
        } catch (e: PrefException) {
            Log.e(TAG, "Preferences error: $e")
        } catch (e: IOException) {
            Log.e(TAG, "IO error: $e")
        } finally {
            conn?.disconnect()
        }
        return false
    }

    companion object {
        private const val TAG = "PlaybackReporter"
    }
}
